pkg/cloudprovider/azure.go

```diff
@@ -51,51 +51,70 @@ type Azure struct {
 	if err != nil {
-		// Fallback to using client ID from env variable if not set.
-		clientID = os.Getenv("AZURE_CLIENT_ID")
-		if strings.TrimSpace(clientID) == "" {
-			return err
-		}
+		klog.Infof("azure_client_id not found in the secret: %v, falling back to AZURE_CLIENT_ID env", err)
+		cfg.clientID = os.Getenv("AZURE_CLIENT_ID")
 	}
-	tenantID, err := a.readSecretData("azure_tenant_id")
+
+	cfg.tenantID, err = a.readSecretData("azure_tenant_id")
 	if err != nil {
-		// Fallback to using tenant ID from env variable if not set.
-		tenantID = os.Getenv("AZURE_TENANT_ID")
-		if strings.TrimSpace(tenantID) == "" {
-			return err
-		}
+		klog.Infof("azure_tenant_id not found in the secret: %v, falling back to AZURE_TENANT_ID env", err)
+		cfg.tenantID = os.Getenv("AZURE_TENANT_ID")
 	}
-	clientSecret, err := a.readSecretData("azure_client_secret")
+
+	cfg.clientSecret, err = a.readSecretData("azure_client_secret")
 	if err != nil {
-		clientSecret = os.Getenv("AZURE_CLIENT_SECRET")
-		// Skip validation; fallback to token workload identity token if env variable is also not set.
-		klog.Infof("Attempting to create workload identity client because azure_client_secret is missing")
+		klog.Infof("azure_client_secret not found in the secret: %v, falling back to AZURE_CLIENT_SECRET env", err)
+		cfg.clientSecret = os.Getenv("AZURE_CLIENT_SECRET")
 	}
-	subscriptionID, err := a.readSecretData("azure_subscription_id")
+
+	cfg.tokenFile, err = a.readSecretData("azure_federated_token_file")
 	if err != nil {
-		return err
+		klog.Infof("azure_federated_token_file not found in the secret: %v, falling back to AZURE_FEDERATED_TOKEN_FILE env", err)
+		cfg.tokenFile = os.Getenv("AZURE_FEDERATED_TOKEN_FILE")
 	}
-	a.resourceGroup, err = a.readSecretData("azure_resourcegroup")
+
+	cfg.subscriptionID, err = a.readSecretData("azure_subscription_id")
+	if err != nil {
+		return nil, fmt.Errorf("azure_subscription_id not found in the secret: %v", err)
+	}
+
+	cfg.resourceGroup, err = a.readSecretData("azure_resourcegroup")
 	if err != nil {
 		if a.platformStatus != nil && len(strings.TrimSpace(a.platformStatus.ResourceGroupName)) > 0 {
 			klog.Infof("Attempting to use resource group from cluster infrastructure because azure_resourcegroup is missing")
-			a.resourceGroup = strings.TrimSpace(a.platformStatus.ResourceGroupName)
+			cfg.resourceGroup = strings.TrimSpace(a.platformStatus.ResourceGroupName)
 		} else {
-			return err
+			return nil, fmt.Errorf("azure_resourcegroup not found in the platform status and the secret: %v", err)
 		}
 	}
-	tokenFile, err := a.readSecretData("azure_federated_token_file")
+
+	return &cfg, nil
+}
+func (a *Azure) initCredentials() error {
+	cfg, err := a.readAzureCredentialsConfig()
 	if err != nil {
-		tokenFile = os.Getenv("AZURE_FEDERATED_TOKEN_FILE")
-		if strings.TrimSpace(tokenFile) == "" {
-			// Use default value if no configuration is set
-			tokenFile = "/var/run/secrets/openshift/serviceaccount/token"
```

Comment:
With the new env var approach this defaulting should not be removed.

Reply:
I defaulted it below. Please let me know if that's where you want that. I could also move it to `readAzureCredentialsConfig`.

Reply:
please move it under https://github.com/openshift/cloud-network-config-controller/pull/151/files#diff-d3dbddd1556958367344cf05b435a98ffdc18ec9c060843fa1a5f4582d22f3b7R90. This is more clear when looking at how this var is set.